File: src/i18n.ts
// Frontend i18n. Strings live in the shared catalog under `locales/<lang>/*.json`
// (the same files the Rust backend embeds), so there is one source of truth per
// language. `t("section.key", { param })` resolves a dotted key and substitutes
// {param} placeholders. Add a language: create locales/<lang>/*.json and register
// it in `catalogs` below.

import enApp from "../locales/en/app.json";
import enClaude from "../locales/en/claude.json";
import enGithub from "../locales/en/github.json";
import enSlack from "../locales/en/slack.json";

type Catalog = Record<string, unknown>;

const catalogs: Record<string, Catalog> = {
  en: { ...enApp, ...enClaude, ...enGithub, ...enSlack },
};

let locale = "en";

// Selectable languages. Add a language: create locales/<id>/*.json, import and
// merge it into `catalogs` above, and add an entry here.
export const LOCALES: { id: string; label: string }[] = [{ id: "en", label: "English" }];

export function availableLocales(): string[] {
  return Object.keys(catalogs);
}

export function getLocale(): string {
  return locale;
}

export function setLocale(next: string): void {
  locale = catalogs[next] ? next : "en";
}

function lookup(cat: Catalog, key: string): unknown {
  return key.split(".").reduce<unknown>((node, seg) => {
    if (node && typeof node === "object") {
      return (node as Record<string, unknown>)[seg];
    }
    return undefined;
  }, cat);
}

export function t(key: string, params?: Record<string, string | number>): string {
  let value = lookup(catalogs[locale] ?? catalogs.en, key);
  if (typeof value !== "string") value = lookup(catalogs.en, key);
  if (typeof value !== "string") return key;

  let out = value;
  if (params) {
    for (const [k, v] of Object.entries(params)) {
      out = out.split(`{${k}}`).join(String(v));
    }
  }
  return out;
}

File: src-tauri/src/engine/i18n.rs
//! Minimal i18n. Strings live in the shared catalog under `locales/<lang>/*.json`
//! (the same files the frontend imports), so there is one source of truth per
//! language. `t(key)` resolves a dotted key (e.g. "claude.tokensByModel") against
//! the catalog; `tf(key, args)` also substitutes `{name}` placeholders. A missing
//! key returns the key itself, so nothing renders blank.
//!
//! Only English is embedded today. Adding a language: drop `locales/<lang>/*.json`
//! and embed + select it here; every caller already goes through `t`/`tf`.

use std::sync::{OnceLock, RwLock};

use serde_json::{Map, Value};

// Selected locale, set once from config at startup. Stored for when more
// languages are embedded; today every lookup resolves against English.
static LOCALE: RwLock<String> = RwLock::new(String::new());

pub fn set_locale(lang: &str) {
    if let Ok(mut current) = LOCALE.write() {
        *current = lang.to_string();
    }
}

fn english() -> &'static Value {
    static CELL: OnceLock<Value> = OnceLock::new();
    CELL.get_or_init(|| {
        // One JSON file per area, merged at the top level.
        const FILES: &[&str] = &[
            include_str!(concat!(
                env!("CARGO_MANIFEST_DIR"),
                "/../locales/en/app.json"
            )),
            include_str!(concat!(
                env!("CARGO_MANIFEST_DIR"),
                "/../locales/en/claude.json"
            )),
            include_str!(concat!(
                env!("CARGO_MANIFEST_DIR"),
                "/../locales/en/github.json"
            )),
        ];
        let mut root = Map::new();
        for raw in FILES {
            if let Ok(Value::Object(map)) = serde_json::from_str::<Value>(raw) {
                for (k, v) in map {
                    root.insert(k, v);
                }
            }
        }
        Value::Object(root)
    })
}

/// Resolve a dotted key. Returns the key itself if it is missing.
pub fn t(key: &str) -> String {
    let mut node = english();
    for seg in key.split('.') {
        match node.get(seg) {
            Some(next) => node = next,
            None => return key.to_string(),
        }
    }
    node.as_str().unwrap_or(key).to_string()
}

/// `t` plus `{name}` placeholder substitution from `args`.
pub fn tf(key: &str, args: &[(&str, &str)]) -> String {
    let mut s = t(key);
    for (name, value) in args {
        s = s.replace(&format!("{{{name}}}"), value);
    }
    s
}
